### ISSUE TITLE
Include C# textmate grammar

### ISSUE THREAD

**Issue by chrisdias** [Microsoft/vscode#19469]
Request from C#/OmniSharp team, grammar is now maintained in an OSS repo:

https://github.com/dotnet/csharp-tmLanguage

**Comment by aeschli** [comment]
So we should add a c# extension again?

**Comment by chrisdias** [comment]
@aeschli I apologize that I don't know the exact workflow around textmate grammar files (see https://github.com/Microsoft/vscode/issues/10481 for proof :).

Is it possible to just pick up the .tmLanguage and not the entire C# extension?  This way we get C# colorization out of the box as we do for other languages (looks like PHP, C++, etc.).

Going forward, the C# folks can maintain the their repo and make a PR to get it into VS Code.

**Comment by aeschli** [comment]
Yes, of course, most of our extensions are just the grammar file and some smart editing rules (brackets, comments...). I'm happy to create one for C#, if that's what the Omnisharp team wants.

**Comment by chrisdias** [comment]
@egamma ok, i see the problem, thanks for the explanation. 

From a customer perspective, having syntax colorization out of the box is goodness.  From an engineering perspective it is a pain to maintain these in multiple places. 

I like the model where the C# grammar is in a single repo that both VS Code and the OmniSharp extension ship:

* the grammar is maintained in a single location
* the grammar is owned by the language owner, not VS Code
* the language owner (or maintainer of the grammar) can make PRs on VS Code to update
* if the VS Code version is out of date, it will be updated when the extension is installed (I assume)
* issues submitted in the VS Code repo can be ported to the grammar repo

Thoughts?

**Comment by egamma** [comment]
@chrisdias see below for comments

>From a customer perspective, having syntax colorization out of the box is goodness. From an engineering perspective it is a pain to maintain these in multiple places.

This is a great summary. However, since we now provide extension recommendations, having syntax colorization out of the box is less important. A user will almost always want to use the rich language extension and not just have coloring that comes with VS Code. 

> the grammar is maintained in a single location
> the grammar is owned by the language owner, not VS Code

This is already the common practice and we should continue to use it, Since the grammars can be used in multiple editors. I must be maintained in a separate repo, e.g., the TypeScript grammar is maintained in https://github.com/Microsoft/TypeScript-TmLanguage. 

>the language owner (or maintainer of the grammar) can make PRs on VS Code to update

This assumes that then VS Code becomes the integrator of grammars. This makes sense for some key languages and can include C#, but it doesn't make sense to me as a general rule. VS Code has extensions.

>if the VS Code version is out of date, it will be updated when the extension is installed (I assume)

This assumes that the grammar is also included in the extension? The version in the extension will win over the version that is bundled with VS Code so this works. However, this means that the grammar is maintained in both VS Code and the extension. This is done in F# today and has resulted in the F# grammar in VS Code being out of date. The F# extension writers only update the grammar in their extension they don't care updating the grammar bundled with VS Code using a PR.

Bottom line we can do this for C#, but not as a general approach.
